addons/hw_l10n_eg_eta/controllers/main.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import base64
+
+import pkcs11
+import platform
+import json
+
+import requests
+from pkcs11 import Attribute, ObjectClass
+from pkcs11.exceptions import NoSuchToken, PinIncorrect, MultipleTokensReturned
+
+from odoo import http
+
+
+class EtaUsbController(http.Controller):
+
+    @http.route('/hw_l10n_eg_eta/certificate', type='http', auth='none', cors='*', csrf=False, save_session=False)
+    def eta_certificate(self, url, otp, user, pin, drive):
+        token, error = self._get_token()
+        if error:
+            return error
+        try:
+            with token.open(user_pin=pin) as session:
+                cert = next(session.get_objects({Attribute.CLASS: ObjectClass.CERTIFICATE}))
+                payload = {
+                    'otp': otp,
+                    'user': user,
+                    'drive': drive,
+                    'certificate': base64.b64encode(cert[Attribute.VALUE]).decode('utf-8')
+                }
+                try:
+                    requests.post(url + '/l10n_eg_edi_eta/certificate', data=payload, timeout=10)
+                except requests.exceptions.Timeout:
+                    return self.timeout_error()
+                except requests.exceptions.ConnectionError:
+                    return self.connection_error()
+
+                return json.dumps({'status': 'success'})
+        except PinIncorrect:
```

Comment:
Why such nesting?  with token.open should only apply until payload={} I suppose, so you could put the outer try..except only around that.  Because PinIncorrect should not be triggered on the rest.  
try: with token except(s)
try: request except(s)